tasks/actorstate/verifreg.go

```diff
@@ -0,0 +1,240 @@
+		if err != nil {
+			// if the actor exists in the current state and not in the parent state then the
+			// actor was created in the current state.
+			if err == types.ErrActorNotFound {
+				return &VerifiedRegistryExtractionContext{
+					PrevState: prevState,
+					CurrState: curState,
+					PrevTs:    a.ParentTipSet,
+					CurrTs:    a.TipSet,
+					Store:     node.Store(),
+				}, nil
+			}
+			return nil, xerrors.Errorf("loading previous verified registry actor at tipset %s epoch %d: %w", a.ParentTipSet.Key(), a.Epoch, err)
+		}
+
+		prevState, err = verifreg.Load(node.Store(), prevActor)
+		if err != nil {
+			return nil, xerrors.Errorf("loading previous verified registry state: %w", err)
+		}
+	}
+	return &VerifiedRegistryExtractionContext{
+		PrevState: prevState,
+		CurrState: curState,
+		PrevTs:    a.ParentTipSet,
+		CurrTs:    a.TipSet,
+		Store:     node.Store(),
+	}, nil
+}
+
+func (VerifiedRegistryExtractor) Extract(ctx context.Context, a ActorInfo, node ActorStateAPI) (model.Persistable, error) {
+	ctx, span := global.Tracer("").Start(ctx, "VerifiedRegistryExtractor")
+	defer span.End()
+
+	stop := metrics.Timer(ctx, metrics.ProcessingDuration)
+	defer stop()
+
+	ec, err := NewVerifiedRegistryExtractorContext(ctx, a, node)
+	if err != nil {
+		return nil, err
+	}
+
+	verifiers, verifierEvents, err := ExtractVerifiers(ctx, ec)
+	if err != nil {
+		return nil, err
+	}
+
+	clients, clientsEvents, err := ExtractVerifiedClients(ctx, ec)
+	if err != nil {
+		return nil, err
+	}
+
+	return model.PersistableList{
+		verifiers,
+		verifierEvents,
+		clients,
+		clientsEvents,
+	}, nil
+}
+
+func ExtractVerifiers(ctx context.Context, ec *VerifiedRegistryExtractionContext) (verifregmodel.VerifiedRegistryVerifiersList, verifregmodel.VerifiedRegistryVerifierEventList, error) {
+	var verifiers verifregmodel.VerifiedRegistryVerifiersList
+	var events verifregmodel.VerifiedRegistryVerifierEventList
+	// if this is the genesis state extract whatever state it has, there is noting to diff against
+	if !ec.HasPreviousState() {
+		if err := ec.CurrState.ForEachVerifier(func(addr address.Address, dcap abi.StoragePower) error {
+			verifiers = append(verifiers, &verifregmodel.VerifiedRegistryVerifier{
+				Height:    int64(ec.CurrTs.Height()),
+				StateRoot: ec.CurrTs.ParentState().String(),
+				Address:   addr.String(),
+				DataCap:   dcap.String(),
+			})
+			events = append(events, &verifregmodel.VerifiedRegistryVerifierEvent{
+				Height:    int64(ec.CurrTs.Height()),
+				StateRoot: ec.CurrTs.ParentState().String(),
+				Event:     verifregmodel.Added,
+				Address:   addr.String(),
+			})
+			return nil
+		}); err != nil {
+			return nil, nil, err
+		}
+		return verifiers, events, nil
+	}
+
+	changes, err := verifreg.DiffVerifiers(ctx, ec.Store, ec.PrevState, ec.CurrState)
+	if err != nil {
+		return nil, nil, xerrors.Errorf("diffing verifier registry verifiers: %w", err)
+	}
+
+	// a new verifier was added
```

Comment:
Do we need separate a separate VerifiedRegistryVerifierEvent model? Can we add the Event field to the VerifiedRegistryVerifier model instead? Add a `MODIFIED` event to represent the datacap being changed.

````Go
type VerifiedRegistryVerifier struct {
	Height    int64  `pg:",pk,notnull,use_zero"`
	StateRoot string `pg:",pk,notnull"`
	Address   string `pg:",pk,notnull"`
        Event string `pg:",pk,notnull,type:verified_registry_event_type"`
	DataCap string `pg:"type:numeric,notnull"`
}
```

When Event = REMOVED then the DataCap field is zero.

Reply:
Good idea -- done.